File: Problems/0121. (E) Best Time to Buy and Sell Stock/Python3/solution.py
class Solution:
  def maxProfit(self, prices: List[int]) -> int:
    # modified Kadane's algorithm (dynamic programming)

    min_price = 10**4
    max_prof = 0
    for p in prices:
      if p < min_price:
        min_price = p
      else:
        max_prof = max(max_prof, p - min_price)
    
    return max_prof

File: Problems/0121. (E) Best Time to Buy and Sell Stock/Java/solution.java
class Solution {
  public int maxProfit(int[] prices) {
    // modified Kadane's algorithm (dynamic programming)

    int min_price = (int) Math.pow(10, 4);
    int max_profit = 0;
    for (int i = 0; i < prices.length; i++) {
      if (prices[i] < min_price)
        min_price = prices[i];
      else
        max_profit = Math.max(max_profit, prices[i] - min_price);
    }
    return max_profit;
  }
}
